### PR TITLE
Make upload_file robust against failures of all kinds

### PR DESCRIPTION
**Pull request by CunningBaldrick**
In particular it now handles failures of complete_upload (I've seen these in the wild).  Also, in put_from_manifest, don't delete backup sstables if they failed to upload (as, unlike snapshots, you can't get them back after deleting them).

### CODE REVIEW

File: cassandra_snapshotter/agent.py [hunk 1]
```diff
@@ -138,16 +160,15 @@ def put_from_manifest(
...
     pool = Pool(concurrency)
-    for ret in pool.imap(upload_file,
-                         ((bucket, f, destination_path(s3_base_path, f), s3_ssenc, buffer_size) for f in files if f)):
-        if not ret:
+    for f in pool.imap(upload_file,
+                       ((bucket, f, destination_path(s3_base_path, f), s3_ssenc, buffer_size) for f in files if f)):
+        if f is None:
+            # Upload failed.
             exit_code = 1
-            break
```
Comment (tbarbugli) [inline cassandra_snapshotter/agent.py]: perhaps I am not reading the diff correctly but, why would you continue the execution if one file failed to upload?

Comment (tbarbugli) [inline cassandra_snapshotter/agent.py]: why wouldn't you want to break as soon as possible here?

File: cassandra_snapshotter/agent.py [hunk 2]
```diff
@@ -138,16 +160,15 @@ def put_from_manifest(
...
+        elif incremental_backups:
+            # Delete files that were successfully uploaded.
+            os.remove(f)
     pool.terminate()
-    if incremental_backups:
```
Comment (tbarbugli) [inline cassandra_snapshotter/agent.py]: perhaps a better approach could be to delete all links only if `exit_code` is 0

**Comment by CunningBaldrick** [comment]
Hi Tommaso, about continuing to upload if one file fails: if one file fails to upload it doesn't necessarily mean that all files will fail to upload.  Since this is a backup tool I think it's best to get files into S3 if it's at all possible.  That way, when your data centre burns down 5 minutes later you will have less work to do recovering from backups, because more stuff will be in S3!  At worst, every file will fail to upload, but who cares?

As for deleting files that successfully uploaded even if other files failed to upload: I guess this might be problematic if the manifest failed to upload but other files successfully uploaded (and were thus deleted), what do you think?  The advantage of deleting files even if other files failed to upload is that you reduce the amount of files you need to upload later when you run the tool again.  Eg if you successfully uploaded 4999 files out of 5000, with one file failing, it's not nice that that one failed file means you have to re-upload the 4999 files next time you run the tool.  However, now that upload is more robust this is less of a consideration.

**Comment by tbarbugli** [comment]
@CunningBaldrick makes sense :)

**Pull request merged**